Enable passing of metamodel params in mm registration API calls

igordejanovic opened the pull request:
In order to make calls `metamodel_for_file/language` symmetric to `metamodel_from_str/file` this adds keywords params to registration API calls that are used to configure the meta-models.


## Code review checklist

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`, no need
      to update for typo fixes and such).


[commit messages]: https://chris.beams.io/posts/git-commit/

igordejanovic committed:
Enable passing of metamodel params in mm registration API calls

igordejanovic committed:
Update CHANGELOG

goto40 commented on docs/registration.md:
```diff
@@ -383,21 +383,31 @@ All classes and functions documented here are directly importable from `textx` m
-- `metamodel_for_language(language_name)` - returns a meta-model for the given language name
+- `metamodel_for_language(language_name, **kwargs)` - returns a meta-model for
+  the given language name. `kwargs` are additional keyword arguments passed to
+  meta-model factory callable, similarly to `metamodel_from_str/file`.
```
Good idea... but I am not sure about the consequences / use case: I just wondered if a language may have some restrictions, some decisions about meta model already made by the language designer. 

--> Is it a use case that I request a metamodel for a (predefined, ready-to-use) language with other meta model parameters that the language designer defined for me...?

igordejanovic replied:
Good point. I found this use case when trying to built meta-model with a particular set of user classes and figured out that a need to override other stuff may be needed. Let's say that those kwargs should be whatever language designer is defined as overridable. Those params are passed to a meta-model callable factory so the factory can decide what to do with those.

Probably some more clarification needed in the docs.

goto40 replied:
This is ok. Since a meta model provider (language designer) can prevent such actions and throw and error if kwargs are passed (and not expected). 

After all, the language factory is fully in the hand of the language designer (or am I wrong?).

For me it is ok.

igordejanovic replied:
Yes, you are right. The language designer can decide what to do with passed parameters.

goto40 commented on docs/registration.md:
```diff
@@ -383,21 +383,31 @@ All classes and functions documented here are directly importable from `textx` m
+    meta-model instances are cached so it is cheap to call
+    `metamodel_for_file/language` multiple times but `kwargs` used in the
+    first call only will be applied as the further calls return cached instance.
+
```
Is it a good idea to *prevent using the cache* when passing some kwargs? 

This would allow to build the customized meta model once (with kwargs), then you can retrieve the registered language afterwards (w/o kwargs).

igordejanovic replied:
Great idea. That would prevent buggy behavior if user would make a call with different kwargs while still retaining good performance in general case. I'll rework it to reinit meta-model whenever kwargs is not empty.

goto40 replied:
Great!

goto40 reviewed (commented):
Nice to hear from you. I have some comments. 

My overall feeling is good. I think the addition is ok. The most critical point is the caching effect, which might lead to hard-to-find bugs in my eyes...

igordejanovic commented:
@goto40 I've reworked the `kwargs` cache circumventing as per your suggestion and updated the docs and added tests. This is ready for another round of review.

goto40 commented:
Good job!
